File: internal/p2p/tagskew_test.go
package p2p

import (
	"crypto/tls"
	"errors"
	"net"
	"strings"
	"testing"
	"time"

	"nib/internal/sign"
)

// P07.S09c: the two skew surfaces D32 had not been given a sentence — the attestation tag, and a
// session protocol with no version in common.

// TestANewerAttestationTagIsLegibleAsASkewRatherThanAsSilence is the fourth surface.
//
// `Attestations` required `[NibCoSign:1]` verbatim, so a signature written by a build that had
// moved to `[NibCoSign:2]` matched nothing and arrived with every field empty. That is
// indistinguishable from a signature that carries no attestation at all — and
// `markOneProceeding` treats an empty commitment on a valid signature as disqualifying, so ONE
// such signature made the whole document report "not one proceeding": an accusation about the
// parties, caused by an upgrade.
func TestANewerAttestationTagIsLegibleAsASkewRatherThanAsSilence(t *testing.T) {
	st := sign.Status{Signers: []sign.SignerInfo{
		{Name: "current", Fingerprint: strings.Repeat("aa", 32), Valid: true,
			Reason: attestationTag + " Accepts X [SPKI:" + strings.Repeat("bb", 32) + "]"},
		{Name: "newer", Fingerprint: strings.Repeat("bb", 32), Valid: true,
			Reason: "[NibCoSign:2] a format this build does not know"},
		{Name: "none", Fingerprint: strings.Repeat("cc", 32), Valid: true,
			Reason: "Finalized in Nib"},
	}}
	got := Attestations(st, Proceeding{})
	if len(got) != 3 {
		t.Fatalf("setup: %d attestations for 3 signatures", len(got))
	}

	// SETUP: this build's own tag still parses, or "the newer one is different" is true of a
	// reader that parses nothing.
	if got[0].TagVersion != attestationTagVersion {
		t.Fatalf("setup: a signature at this build's own tag version reports TagVersion %d, want %d",
			got[0].TagVersion, attestationTagVersion)
	}
	if got[0].AcceptedPeer == "" {
		t.Fatal("setup: this build no longer parses its own attestation format")
	}

	// The newer one: legible AS a skew.
	if got[1].TagVersion != 2 {
		t.Errorf("a signature tagged [NibCoSign:2] reports TagVersion %d. Without the version a "+
			"reader cannot tell 'this build cannot read that signature' from 'that party "+
			"attested to nothing', and the second is an accusation", got[1].TagVersion)
	}
	// And still not TRUSTED: a format this build does not know has no fields it may interpret.
	if got[1].AcceptedPeer != "" || got[1].RosterHash != "" {
		t.Errorf("a newer tag's fields were parsed anyway (peer=%q roster=%q). Reading the "+
			"version is not the same as understanding the payload, and guessing at a format this "+
			"build does not know is how a skew becomes a wrong answer rather than no answer",
			got[1].AcceptedPeer, got[1].RosterHash)
	}

	// A signature with NO tag is the third state and must stay distinct from both.
	if got[2].TagVersion != 0 {
		t.Errorf("an ordinary Finalize signature reports TagVersion %d, want 0 — a signature that "+
			"is not a Nib attestation at all is not a version skew", got[2].TagVersion)
	}
}

// TestADisjointProtocolIsASentenceNotATLSAlert is the third surface, and it drives a REAL
// handshake rather than asserting the substring the classifier matches on.
//
// `asProtocolSkew` matches "no application protocol" in somebody else's error text. A test that
// asserted that constant would be checking the constant against itself; this one stands up a
// listener speaking a protocol this build does not offer and dials it for real, so if the stdlib
// ever rewords the alert this goes red instead of silently returning the raw error.
func TestADisjointProtocolIsASentenceNotATLSAlert(t *testing.T) {
	certPEM, keyPEM, err := sign.GenerateIdentity("listener")
	if err != nil {
		t.Fatal(err)
	}
	fp, err := sign.Fingerprint(certPEM)
	if err != nil {
		t.Fatal(err)
	}
	dialCert, dialKey, err := sign.GenerateIdentity("dialer")
	if err != nil {
		t.Fatal(err)
	}
	dialFP, err := sign.Fingerprint(dialCert)
	if err != nil {
		t.Fatal(err)
	}
	srvCfg, err := SessionTLS(certPEM, keyPEM, dialFP, true)
	if err != nil {
		t.Fatal(err)
	}
	// A build from the future: it speaks only a version this one has never heard of.
	srvCfg.NextProtos = []string{"nib/99"}
	srvCfg.ClientAuth = tls.NoClientCert
	ln, err := tls.Listen("tcp", "127.0.0.1:0", srvCfg)
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()
	go func() {
		for {
			c, aerr := ln.Accept()
			if aerr != nil {
				return
			}
			_ = c.(*tls.Conn).Handshake()
			c.Close()
		}
	}()

	_, derr := Dial(t.Context(), ln.Addr().String(), dialCert, dialKey, fp, 3*time.Second)
	if derr == nil {
		t.Fatal("dialling a peer that speaks no protocol version in common succeeded")
	}

	var skew *ProtocolSkewError
	if !errors.As(derr, &skew) {
		t.Fatalf("a disjoint protocol arrived as %v — a raw TLS alert on the connect path, where "+
			"every other failure means the peer is unreachable. D32: a version skew produces a "+
			"sentence naming the mismatch, not a parse error, and an alert IS the parse error "+
			"one layer down", derr)
	}
	msg := skew.Error()
	if !strings.Contains(msg, "version of Nib") || !strings.Contains(msg, "Update") {
		t.Errorf("the sentence does not name the mismatch or the fix: %q", msg)
	}
	// It says which versions were on the table, so a user reporting it says something useful.
	for _, want := range sessionALPN {
		if !strings.Contains(msg, want) {
			t.Errorf("the sentence does not name this build's offer %q: %q", want, msg)
		}
	}
}

// TestAnOrdinaryDialFailureIsNotCalledAVersionSkew is the control. A classifier that returned the
// typed error for every handshake failure would relabel every unreachable peer as an out-of-date
// one — which is the same misdiagnosis in the other direction, and the more common case by far.
func TestAnOrdinaryDialFailureIsNotCalledAVersionSkew(t *testing.T) {
	// A closed port: the ordinary "nothing is listening" failure.
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	addr := l.Addr().String()
	l.Close()

	certPEM, keyPEM, err := sign.GenerateIdentity("dialer")
	if err != nil {
		t.Fatal(err)
	}
	_, derr := Dial(t.Context(), addr, certPEM, keyPEM, make([]byte, 32), 3*time.Second)
	if derr == nil {
		t.Fatal("setup: dialling a closed port succeeded")
	}
	var skew *ProtocolSkewError
	if errors.As(derr, &skew) {
		t.Errorf("an unreachable peer was reported as a version skew: %v. Telling a user to "+
			"update Nib when the other machine is simply not listening sends them to the one "+
			"place the problem is not", derr)
	}
}

File: test/jsdom/tagskew.test.mjs
// A signature this build cannot READ is not a signature that disagrees (P07.S09c, D32, C13).
//
// ── The fourth skew surface, and the one that produced an accusation ────────
// D32 says a version skew "produces a sentence naming the mismatch, not a parse error". Three
// surfaces had one: the ceremony record, the invitation, and the session protocol. The fourth —
// the attestation tag inside every signature's /Reason — was excused, and its skew was the worst
// of the four because it did not fail loudly at all.
//
// `Attestations` required `[NibCoSign:1]` verbatim before it would read anything, so a signature
// from a build that had moved to `[NibCoSign:2]` matched nothing: no acceptedPeer, no rosterHash.
// `markOneProceeding` treats an empty commitment on a VALID signature as disqualifying, so ONE
// such signature made the whole document report *"This document was not produced by a single
// agreed proceeding."*
//
// An accusation about the parties, on a document everybody signed correctly, caused by one of
// them having updated Nib — which is verbatim the harm the roster token's own version was added
// to prevent one level down.
import test from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import { boot, REPO } from './boot.mjs';
import { setNextDocument } from './stub-pdfjs.mjs';

const A = 'aa'.repeat(32);
const B = 'bb'.repeat(32);
const ROSTER = 'c0ffee'.repeat(10) + 'abcd';

// att builds one attestation as `/api/attestations` serialises it. `tagVersion` is the field
// under test; 1 is what this build writes.
const att = (fp, peer, over = {}) => ({
  signer: 'S-' + fp.slice(0, 2), fingerprint: fp, acceptedPeer: peer,
  reason: '[NibCoSign:1] Accepts [SPKI:' + peer + ']', when: '2026-08-21T10:00:00Z',
  valid: true, matched: true, pinned: false, rosterHash: ROSTER, rosterVersion: 4,
  oneProceeding: true, tagVersion: 1,
  ...over,
});

// A signature from a NEWER build: Go read the version and deliberately parsed nothing else, so
// every token field is empty and only `tagVersion` says why.
const fromNewerNib = (fp) => ({
  signer: 'S-' + fp.slice(0, 2), fingerprint: fp, acceptedPeer: '',
  reason: '[NibCoSign:2] something this build does not know how to read',
  when: '2026-08-21T10:02:00Z', valid: true, matched: false, pinned: false,
  rosterHash: '', rosterVersion: 0, oneProceeding: false, tagVersion: 2,
});

let attestations = [];
const { document: doc, settle } = await boot({
  routes: {
    '/api/attestations': () => ({ attestations }),
    '/api/open': {
      id: 'test-epoch:1', name: 'deed.pdf', path: '/tmp/nib-harness/deed.pdf',
      canSave: false, canUndo: false, canRedo: false,
      signature: {
        state: 'valid',
        signers: [
          { name: 'S-aa', valid: true, when: '2026-08-21T10:00:00Z' },
          { name: 'S-bb', valid: true, when: '2026-08-21T10:01:00Z' },
        ],
      },
    },
  },
});

setNextDocument({ numPages: 1 });
doc.getElementById('pathInput').value = '/tmp/nib-harness/deed.pdf';
doc.getElementById('openGo').click();
await settle();

async function panelText(next) {
  attestations = next;
  const body = doc.getElementById('sigDetailsBody');
  body.innerHTML = '';
  doc.getElementById('sigDetailsBtn').click();
  // Wait for the ATTESTATION box, not for any text: the signer rows render synchronously and
  // `augmentSigDetails` is a fetch behind them, so waiting on `textContent` returns a panel that
  // has drawn the rows and none of the verdicts — which reads as "the fix does nothing".
  for (let i = 0; i < 80 && !body.querySelector('.sigatt'); i++) {
    await new Promise((r) => setTimeout(r, 5));
  }
  return body.textContent.replace(/\s+/g, ' ');
}

test('a signature from a newer Nib is reported as unreadable, not as a disagreement', async () => {
  const txt = await panelText([att(A, B), fromNewerNib(B)]);

  // STIMULUS: the panel drew. Without this the absence assertion below passes against a panel
  // that rendered nothing, which is what a broken fetch produces.
  assert.ok(txt.length > 0, 'the panel rendered nothing, so nothing below measures anything');

  assert.ok(!/was not produced by a single agreed proceeding/.test(txt),
    `a document whose only anomaly is that one party has a NEWER Nib was reported as not being ` +
    `one proceeding. That is an accusation about the parties caused by an upgrade — verbatim the ` +
    `harm D32 exists to prevent, and the harm the roster token's own version already prevents ` +
    `one level down: ${txt}`);
  assert.match(txt, /newer version of Nib and this one cannot read them/,
    `the panel says nothing about the skew, so a reader is left with a document that has one ` +
    `signature carrying no attestation and no explanation of why: ${txt}`);
  assert.match(txt, /version difference, not a disagreement/,
    `the sentence does not distinguish the two, which is the whole point of saying it: ${txt}`);
});

test('a document all of whose signatures this build can read still gets its verdict', async () => {
  // The control, and it is what the fix would break if the new branch swallowed the old one.
  // Suppressing the proceeding verdict whenever ANY tagVersion is present — rather than a newer
  // one — would silence the check on every ordinary ceremony in the product.
  const txt = await panelText([att(A, B), att(B, A)]);
  assert.match(txt, /✓ One proceeding/,
    `a ceremony this build can read in full lost its proceeding verdict: ${txt}`);
  assert.ok(!/cannot read them/.test(txt),
    `signatures at this build's own tag version were reported as unreadable: ${txt}`);
});

test('a genuine disagreement is still reported when every signature is readable', async () => {
  // The other control. The unreadable branch stands in FRONT of the disagreement line, so a
  // version that suppressed it unconditionally would hide real disagreements — the failure
  // `oneproceeding.test.mjs` was written against, reintroduced by its own fix.
  const other = 'dec0de'.repeat(10) + 'abcd';
  const txt = await panelText([att(A, B), att(B, A, { rosterHash: other, oneProceeding: false })]);
  assert.match(txt, /⚠ Not one proceeding/,
    `two signatures naming DIFFERENT ceremonies were no longer reported: ${txt}`);
});

// TestTheClientAndGoAgreeOnTheTagVersion — the cross-language constant.
//
// The number lives in two languages: `attestationTagVersion` in Go decides what gets parsed, and
// `ATTESTATION_TAG_VERSION` in app.js decides what gets reported as unreadable. If Go moves to 2
// and the client does not, every signature this build writes reports itself as unreadable to its
// own user — the skew sentence firing on a document with no skew in it.
test('the client and Go agree on the attestation tag version', () => {
  const goSrc = fs.readFileSync(path.join(REPO, 'internal/p2p/attestation.go'), 'utf8');
  const jsSrc = fs.readFileSync(path.join(REPO, 'web/app.js'), 'utf8');

  const go = goSrc.match(/attestationTagVersion\s*=\s*(\d+)/);
  const js = jsSrc.match(/ATTESTATION_TAG_VERSION\s*=\s*(\d+)/);
  assert.ok(go, 'attestationTagVersion is not declared in Go; this guard read the wrong file');
  assert.ok(js, 'ATTESTATION_TAG_VERSION is not declared in app.js');
  assert.equal(js[1], go[1],
    `app.js reads attestation tag version ${js[1]} and Go writes ${go[1]}. A build whose two ` +
    `halves disagree reports its OWN signatures as unreadable to its own user — the skew ` +
    `sentence firing on a document with no skew in it.`);

  // And the tag constant itself carries that version, or the parser and the writer disagree about
  // what a current signature looks like.
  assert.ok(goSrc.includes('attestationTag = "[NibCoSign:' + go[1] + ']"'),
    `attestationTag does not carry version ${go[1]}, so what this build WRITES and what it ` +
    `considers current are two different numbers`);
});
